### PR TITLE
fix a couple of ESS issues during ES/ESS restart

### PR DESCRIPTION
**Pull request by birarda**
- fixed a bug where the AudioScriptingInterface of the Entity Script Server would be locked up if the Entity Server restarted
- fixed multiple bugs where the Entity Script Server would crash if the Entity Server restarted

### CODE REVIEW

**Comment by birarda** [comment]
#### Test Plan

1. Run the Sandbox from this PR.
2. Spawn Shortbow in the grid of your sandbox by finding it in the marketplace.
3. Restart your Entity Server (by clicking the X on for that node on your domain-server nodes page).
4. Wait 10s.
5. Click the start button for Shortbow. The game should start and work as expected.
6. Repeat the above steps with master. The game should not function on step 5.

**Comment by ZappoMan** [comment]
@birarda - please add this test plan to the test plan library wiki

**Comment by hifi-gustavo** [comment]
Build console output for commit: 713eb1ace2e7977623ac36d2d5fc903d621ce061 available at: https://gist.github.com/fe610dc7ae96f9da06a21af589c12176<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9922-713eb1a.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9922-713eb1a.dmg

**Comment by ZappoMan** [comment]
@birarda - Since this changes the implementation of the client's Audio Scripting interface to use dependencies, we should also rerun the test plans from the test library that read on that feature.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 684da724fee9e64492be4b0045f3033499892706 available at: https://gist.github.com/fe610dc7ae96f9da06a21af589c12176<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9922-684da72.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9922-684da72.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: ce02184cdd6ebfad01291a8289e4d08b510ce1b0 available at: https://gist.github.com/fe610dc7ae96f9da06a21af589c12176<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9922-ce02184.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9922-ce02184.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: a8d9519002f5c93443d6b863cec437d085c16efc available at: https://gist.github.com/fe610dc7ae96f9da06a21af589c12176<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9922-a8d9519.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9922-a8d9519.dmg

**Comment by Nex-Pro** [comment]
Testing this.

**Comment by Nex-Pro** [comment]
Run the Sandbox from this PR. **Done**
Spawn Shortbow in the grid of your sandbox by finding it in the marketplace. **Done**
Restart your Entity Server (by clicking the X on for that node on your domain-server nodes page).
Wait 10s.**Done**
Click the start button for Shortbow. The game should start and work as expected.**Fail**
***Could only get the game to start by restarting the entity script server.***

Repeat the above steps with master. The game should not function on step 5. did not compare as this pr failed.

debugWindow.js Test

Signals used by debugWindow.js have been changed with this PR. While scripts are running open the HMD script debug window and ensure it is receiving output from the scripts. **SUCCESS**

You can open this window by running the script at developer > debugging > debugWindow.js or via the menu item Developer > Script Log (HMD Friendly). **done**

NOISE-GATE-03 Test
Verify the functionality of the AudioScriptingInterface (which is now a dependency) using the following test on this PR.

Objective: Verify that reports of noise gate opening and closing are always in sequential pairs, that we don't report that the noise gate closes more than once in a row.
Precondition: Enter any domain, make sure that the domain is totally silent other than your own client. Make sure the Audio Noise Reduction is enabled.

Run interface.
Open Edit > Console.
Enter the following commands:
Audio.noiseGateOpened.connect(function() { print("opened"); }); Audio.noiseGateClosed.connect(function() { print("closed"); });
Verify that in the log you see "opened" followed by "closed" and then repeated each time the noise gate opens or closes. **SUCCESS**


@birarda   Failed part one.

**Comment by Nex-Pro** [comment]
Tried multiple times

**Comment by Nex-Pro** [comment]
Retesting in progress.

**Comment by birarda** [comment]
Can't edit a comment on mobile. For the original test plan to avoid conflating other bugs that will be fixed outside this PR please kill the ES in step 3 by identifying it in the task manager and killing the process.

**Comment by Nex-Pro** [comment]
### PASS

 Pass following new testing notes from @birarda  Killed ent server via task manager.
Worked on each attempt.

File: assignment-client/src/scripts/EntityScriptServer.cpp [hunk 1]
```diff
@@ -324,7 +326,26 @@ void EntityScriptServer::nodeActivated(SharedNodePointer activatedNode) {
...
         case NodeType::EntityServer: {
-            clear();
+            // Before we clear, make sure this was our only entity server.
+            // Otherwise we're assuming that we have "trading" entity servers
+            // (an old one going away and a new one coming onboard)
+            // and that we shouldn't clear here because we're still doing work.
+            bool hasAnotherEntityServer = false;
+            auto nodeList = DependencyManager::get<NodeList>();
+
+            nodeList->eachNodeBreakable([&hasAnotherEntityServer, &killedNode](const SharedNodePointer& node){
```
Comment (ZappoMan) [inline assignment-client/src/scripts/EntityScriptServer.cpp]: not a blocker - but wouldn't it be easier to just keep track of all the `nodeAdded` and `nodeKilled` calls? And just keep a counter?

**Comment by hifi-gustavo** [comment]
Build console output for commit: 03170718006ac79a2d16f2cceb6626a4bf9cdce5 available at: https://gist.github.com/fe610dc7ae96f9da06a21af589c12176<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9922-0317071.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9922-0317071.dmg

**Comment by birarda** [comment]
@Nex-Pro if you can re-test this using the original modified test plan for Shortbow, that would be excellent.

**Comment by Nex-Pro** [comment]
Restesting with original notes.

**Comment by Nex-Pro** [comment]
Had to drop this after discovering an issue with the shortbow game not being delivered correctly. Had installed, uninstalled and cleaned this several times.

**Comment by birarda** [comment]
@Nex-Pro I'll look into that and let you know when this is good to test again.

**Comment by birarda** [comment]
@Nex-Pro I have fixed the issues with the marketplace version of shortbow and it should be spawnable now - please re-test this PR with that when you get a chance.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 96cf6b8c0418ca4e8548792d921405d70c81db30 available at: https://gist.github.com/fe610dc7ae96f9da06a21af589c12176<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9922-96cf6b8.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9922-96cf6b8.dmg

**Comment by Nex-Pro** [comment]
Will continue this test shortly.

**Comment by Nex-Pro** [comment]
Testing this now.

**Comment by Nex-Pro** [comment]
**Testing platform**
**Intel Workstation**
**CPU I7 4.0ghz.**
**Ram: 16 gb.**
**GPU: Nvidia Gtx 1070 8gb Vram**
**HMD: HTV Vive**

IMO ** PASS **  Please read notes.

### Test Plan

## Shortbow Test

Run the Sandbox from this PR.**: DONE**

Spawn Shortbow in the grid of your sandbox by finding it in the marketplace.**: DONE**

Restart your Entity Server (by clicking the X on for that node on your domain-server nodes page).**: DONE**

Wait 10s.**: DONE**

Click the start button for Shortbow. The game should start and work as expected.**: SUCCESS** Multiple successful passes.

Repeat the above steps with master. The game should not function on step 5.**:CONFIRMED**


## debugWindow.js Test


Signals used by debugWindow.js have been changed with this PR.

 While scripts are running open the HMD script debug window and ensure it is receiving output from the scripts.**: SUCCESS**

You can open this window by running the script at developer > debugging > debugWindow.js or via the menu item Developer > Script Log (HMD Friendly).**: DONE**



## NOISE-GATE-03 Test

Verify the functionality of the AudioScriptingInterface (which is now a dependency) using the following test on this PR.

Objective: Verify that reports of noise gate opening and closing are always in sequential pairs, that we don't report that the noise gate closes more than once in a row.

Precondition: Enter any domain, make sure that the domain is totally silent other than your own client. Make sure the Audio Noise Reduction is enabled.**: CHECK**

Run interface.

Open Edit > Console.**: DONE**

Enter the following commands:

Audio.noiseGateOpened.connect(function() { print("opened"); });**: DONE**

Audio.noiseGateClosed.connect(function() { print("closed"); });**: DONE**

Verify that in the log you see "opened" followed by "closed"**: SUCCESS**

and then repeated each time the noise gate opens or closes.**: SUCCESS**


## Smoke Test:

## QA Test
**QA Tester**: Piper.Peppercorn
**localhost 'createTestBlocks.js' Result**: SUCCESS
**localhost ATP Upload + Edit Result**: SUCCESS
**Echo Server Audio**: SUCCESS
**localhost 'harmonicOscillator.js' Result**: SUCCESS
**localhost 'reverbTest.js' Result**: SUCCESS
**localhost 'playaPerformanceTest.js' Result**: SUCCESS
**dev-playa 'playaPerformanceTest.js' Result**: SUCCESS
**dev-demo 'playaPerformanceTest.js' Result**: SUCCESS
**QA Pass**: SUCCESS

**Additional Test Results + Observations**: Had a situation during smoke test when I was not and could not connect to localhost until restarting interface. See log.


**Domain** |localhost |dev-playa |
---|---|---|---
**Render Rate**: |90 |89 |
**Present Rate**: |90 |90 |
**Present New Rate**: |90 |90 |
**CPU Buffer Count**: |2302 |3746 |
**GPU Buffer Count**: |869 |885 |
**CPU Memory Usage**: |24954896 |66273584 |
**GPU Memory Usage**: |17335600 |36937776 |
**CPU Texture Count**: |261 |261 |
**GPU Texture Count**: |214 |181 |
**GPU Transfer Count**: |0 |0 |
**CPU Texture Memory Usage**: |85169672 |272297440 |
**GPU Texture Memory Usage**: |1050060362 |1009224456 |
**GPU Virtual Memory Usage**: |1050060362|1009224456|
**Triangle Count**: |634538 |2229591 |
**Triangle Rate**: |57685272 |222959100 |
**API Drawcall Count**: |101 |354 |
**Drawcall Count**: |198 |978 |
**Drawcall Rate**: |18000 |97800 |
**State Changes Texture Count**:|121 |316 |
**Pipeline Count**: |45 |44 |
**Item Opaques**: |112 |517 |
**Item Translucents**: |31 |73 |
**Item Lights**: |4 |50 |
**Timing Opaques**: |0.513 |1.814 |
**Timing Translucents**: |0.045 |0.111 |
**Timing Lighting**: |0.011 |0.029 |
**Render Frame**: |1.068 |3.585 |

[hifi-log.zip](https://github.com/highfidelity/hifi/files/859287/hifi-log.zip)

**Pull request merged**